From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <nonplay@bxteam.org>
Date: Sun, 19 Jul 2026 00:34:38 +0300
Subject: [PATCH] Raytrace Entity Culling

Server-side port of tr7zw's EntityCulling mod.
Entities that a player provably cannot see (fully hidden behind opaque blocks)
are removed from that player's entity tracker. This reduces network bandwidth,
improves client FPS and blinds entity-ESP cheats (e.g. freecam/x-ray style hacks
cannot see mobs, players or items through walls anymore).

diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index 7ed727cbb33e490257283400220ad6a2a3bc9918..e482645097fdbd6b0994b839cd0d5b56cd5e12e6 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -1555,6 +1555,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                 && this.entity.broadcastToPlayer(player)
                 && ChunkMap.this.isChunkTracked(player, this.entity.chunkPosition().x(), this.entity.chunkPosition().z());
             // Paper end - Configurable entity tracking range by Y
+            visibleToPlayer = visibleToPlayer && !org.bxteam.divinemc.culling.EntityCullingManager.isCulled(player, this.entity); // DivineMC - Raytrace Entity Culling
             // CraftBukkit start - respect vanish API
             if (visibleToPlayer && !player.getBukkitEntity().canSeeChunkMapUpdatePlayer(this.entity.getBukkitEntity())) { // Paper - only consider hits // DivineMC - SparklyPaper: Optimize "canSee" checks
                 visibleToPlayer = false;
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 0c7a74b1781a62d8aa700ea9255ffae66daf9a04..149aa10058f16a45e58ac4db5493340264c78c5e 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -444,6 +444,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     public boolean suppressTrackerForLogin = false; // Paper - Fire PlayerJoinEvent when Player is actually ready
     // CraftBukkit end
     public boolean isRealPlayer; // Paper
+    public volatile org.bxteam.divinemc.culling.@Nullable PlayerCullTracker cullTracker; // DivineMC - Raytrace Entity Culling
     public com.destroystokyo.paper.event.entity.@Nullable PlayerNaturallySpawnCreaturesEvent playerNaturallySpawnedEvent; // Paper - PlayerNaturallySpawnCreaturesEvent
     public org.bukkit.event.player.PlayerQuitEvent.@Nullable QuitReason quitReason = null; // Paper - Add API for quit reason; there are a lot of changes to do if we change all methods leading to the event
     private boolean tpsBar = false; // Purpur - Implement TPSBar
@@ -834,6 +835,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
 
     public void doTick() {
         try {
+            org.bxteam.divinemc.culling.EntityCullingManager.tickPlayer(this); // DivineMC - Raytrace Entity Culling
             if (this.valid && (!this.isSpectator() || !this.touchingUnloadedChunk())) { // Paper - don't tick dead players that are not in the world currently (pending respawn)
                 super.tick();
                 if (!this.containerMenu.stillValid(this)) {
diff --git a/net/minecraft/world/entity/EntityType.java b/net/minecraft/world/entity/EntityType.java
index 8f0893d42dd56d8a3cac5d142f7d2b0d23eb435c..78d0c7af20b1fabc3994d34dab685a42d8e332a4 100644
--- a/net/minecraft/world/entity/EntityType.java
+++ b/net/minecraft/world/entity/EntityType.java
@@ -68,6 +68,7 @@ public class EntityType<T extends Entity> implements EntityTypeTest<Entity, T>,
     private final int clientTrackingRange;
     private final int updateInterval;
     public boolean dabEnabled = false; // DivineMC - Dynamic Activation of Brain
+    public boolean raytraceCullingSkip = false; // DivineMC - Raytrace Entity Culling
     private final String descriptionId;
     private @Nullable Component description;
     private final Optional<ResourceKey<LootTable>> lootTable;
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 39198a4de8ca0d59211b80ef588b6b433ea28666..b30ffaaa38ddf1527ffb87084b6565a8a7c0d72d 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1162,6 +1162,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         }
         // DivineMC end - Imanity: Raytrace AntiXraySDK integration
         this.chunkPacketBlockController.onBlockChange(this, pos, blockState, oldState, updateFlags, updateLimit); // Paper - Anti-Xray
+        org.bxteam.divinemc.culling.EntityCullingManager.onBlockChange(this, pos); // DivineMC - Raytrace Entity Culling
         if (oldState == null) {
             // CraftBukkit start - remove blockstate if failed (or the same)
             if (captureBlockStates && captured) { // DivineMC - Regionized Chunk Ticking - thread-safe block capture
diff --git a/net/minecraft/world/waypoints/WaypointTransmitter.java b/net/minecraft/world/waypoints/WaypointTransmitter.java
index 07af92310ee5c8bba04e6175b6faac629632e4de..5f3059c4342b59664fa2da97832815fa5024391e 100644
--- a/net/minecraft/world/waypoints/WaypointTransmitter.java
+++ b/net/minecraft/world/waypoints/WaypointTransmitter.java
@@ -21,6 +21,7 @@ public interface WaypointTransmitter extends Waypoint {
 
     static boolean doesSourceIgnoreReceiver(final LivingEntity source, final ServerPlayer receiver) {
         if (!receiver.getBukkitEntity().canSee(source.getBukkitEntity())) return true; // Paper - ignore if entity is hidden from player
+        if (org.bxteam.divinemc.culling.EntityCullingManager.isCulled(receiver, source)) return true; // DivineMC - Raytrace Entity Culling - hide culled players from locator bar
         if (receiver.isSpectator()) {
             return false;
         } else if (!source.isSpectator() && !source.hasIndirectPassenger(receiver)) {
